x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/parser/ExpressionBuilder.java

```diff
@@ -721,13 +721,25 @@ public Literal visitParamLiteral(ParamLiteralContext ctx) {
         }
         // otherwise we need to make sure that xcontent-serialized value is converted to the correct type
         try {
-            if (canConvert(sourceType, dataType) == false) {
+            Object literalValue;
+            // The interval types are handled "directly", as they must not be exposed through the SQL types converter, which would then in
+            // turn expose the "protocol" interval types (i.e. "interval_year" & co.) to, for instance, SQL scalars (like CAST) a.s.o.
+            if (SqlDataTypes.isYearMonthInterval(dataType)) {
```

Comment:
I think this logic could be encapsulated into `Intervals` class:

```java
if (SqlDataTypes.isInterval(dataType)) {
    return Intervals.from(param.value, dataType);
}
```

Reply:
This would look better there indeed! Pushed an update, hope it matches your suggestion.